### PR TITLE
Refactor np.interp

### PR DESCRIPTION
**Pull request by rjenc29**
Initial commit; a long way to go.

Ambition is to replicate `np.interp` exactly for all versions 1.10+ by straight porting the underlying numpy implementation and its changes through time / versions.

Will update this PR as it progresses towards that goal and invite a review when it looks to be in good shape.

### CODE REVIEW

**Comment by rjenc29** [comment]
**To recap** 
The intent of this change is to replicate numpy as closely as possible for all versions 1.10+ by directly porting the underlying C implementation.

This is in an attempt to provide stronger guarantees in respect of consistency of output.

As with the numpy implementation, I have implemented separate (but very similar) real and complex valued algorithms - and I have done so with pre- and post-116 flavours, with a numpy version gate to decide which one(s) to point at in the `overload`.

Correspondingly, there's a lot of apparently repetitious code.  This could be trivially factored out, but probably not without making the mapping back to numpy source more difficult.  As there are some subtle and sometimes unexpected behaviours in `np.interp` I chose to make the mapping as obvious as possible in the first instance, but I am open to suggestion.

There's a change in master which I added to the real values implementations but commented out as I am not sure if it's in any 1.16.x version yet (I could be wrong).  I left a comment in the complex valued algos.

I have put in extra tests - particularly around complex; non-finite and non-monotonic inputs. 

If you get a mo, please could you review and let me know your thoughts.

TA!

File: numba/tests/test_np_functions.py [hunk 1]
```diff
@@ -2102,139 +2103,134 @@ def test_interp_basic(self):
...
-            with self.assertRaises(ValueError) as e:
-                cfunc(x, xp, fp)
-
-            self.assertIn(msg, str(e.exception))
-
-        x = np.arange(6)
-        xp = np.array([1, 2, 3, 3, 3, 5])  # repeating values
-        fp = np.arange(6)
-        _check(x, xp, fp)
+        ndata = 20000
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: need this value be so large?

**Comment by stuartarchibald** [comment]
@rjenc29 I made fixes here: https://github.com/rjenc29/numba/pull/1 given they are all just commentary, and one value change to make the tests run quicker, if you are happy please merge and this branch will get an update to address the comments.

**Comment by rjenc29** [comment]
All changes look good to me, thanks!

**Pull request merged**